### PR TITLE
adding getMultiple API support and unit test

### PR DESCRIPTION
**Pull request by Caseyyu2**
adding getMultiple API support and unit test

### CODE REVIEW

File: src/tests/NoSqlProviderTests.ts [hunk 1]
```diff
@@ -2419,7 +2424,7 @@ describe('NoSqlProvider', function () {
             }
 
             it('Full Text Index', (done) => {
-                return openProvider(provName, {
```
Comment (masterkidan) [inline src/tests/NoSqlProviderTests.ts:2422]: why these changes?

File: src/tests/NoSqlProviderTests.ts [hunk 2]
```diff
@@ -858,7 +858,12 @@ describe('NoSqlProvider', function () {
                                         assert.equal(ret.length, 2);
                                         ret.forEach(r => { assert.equal(r.val, 'b'); });
                                     });
-                                    return Promise.all([g1, g2, g2b, g2c, g3, g4]).then(() => {
+                                    var g5 = prov.getMultiple('test', ['x', 'y'], 'key').then(retVal => {
```
Comment (masterkidan) [inline src/tests/NoSqlProviderTests.ts:861]: Can you make this a separate test?

File: src/InMemoryProvider.ts [hunk 3]
```diff
@@ -369,6 +369,22 @@ class InMemoryIndex extends DbIndexFTSFromRangeQueries {
...
+        if (isError(joinedKeys)) {
+            return Promise.reject(joinedKeys);
+        }
+        
+        if (Array.isArray(keyOrKeys)) {
+            var sortedKeys = keyOrKeys.sort();
+            return this.getRange(sortedKeys[0], sortedKeys[sortedKeys.length - 1], false, false);
+        }
+
+        return this.getRange(keyOrKeys, keyOrKeys, false, false);
```
Comment (masterkidan) [inline src/InMemoryProvider.ts:385]: You should probably do a pick after this, to only pick the items that match the keys. I imagine this would return more than just the list of keys

Comment (masterkidan) [reply src/InMemoryProvider.ts:385]: It may be better to just update this to the same logic as indexeddb ... .i.e make n get() calls

File: src/NoSqlProvider.ts [hunk 4]
```diff
@@ -169,7 +170,15 @@ export abstract class DbProvider {
...
-    getMultiple(storeName: string, keyOrKeys: KeyType | KeyType[]): Promise<ItemType[]> {
+    getMultiple(storeName: string, keyOrKeys: KeyType | KeyType[], indexName?: string): Promise<ItemType[]> {
+
+        if (indexName) {
+            return this._getStoreIndexTransaction(storeName, false, indexName).then(index => {
+                // only inMemory Provider doesn't have the support
+                return index.getMultiple(keyOrKeys);
+            });
+        }
+        // in worker scenario we only issue request for indexedDBProvider
```
Comment (masterkidan) [inline src/NoSqlProvider.ts:181]: Please remove this comment

File: src/NoSqlProvider.ts [hunk 5]
```diff
@@ -169,7 +170,15 @@ export abstract class DbProvider {
         });
     }
 
-    getMultiple(storeName: string, keyOrKeys: KeyType | KeyType[]): Promise<ItemType[]> {
+    getMultiple(storeName: string, keyOrKeys: KeyType | KeyType[], indexName?: string): Promise<ItemType[]> {
+
+        if (indexName) {
+            return this._getStoreIndexTransaction(storeName, false, indexName).then(index => {
+                // only inMemory Provider doesn't have the support
```
Comment (masterkidan) [inline src/NoSqlProvider.ts:177]: I thought you just added support, why doesn't it support?

File: src/tests/NoSqlProviderTests.ts (hunk 1 above)
Comment (Caseyyu2) [reply src/tests/NoSqlProviderTests.ts:2422]: This is due to In Mocha v3.0.0 and newer, returning a Promise and calling done() will result in an exception. Without those change, the unit tests don't run correctly.

See more in this post https://github.com/mochajs/mocha/issues/2407

File: src/NoSqlProvider.ts (hunk 5 above)
Comment (Caseyyu2) [reply src/NoSqlProvider.ts:177]: removed

File: src/NoSqlProvider.ts (hunk 4 above)
Comment (Caseyyu2) [reply src/NoSqlProvider.ts:181]: removed

File: src/tests/NoSqlProviderTests.ts (hunk 1 above)
Comment (masterkidan) [reply src/tests/NoSqlProviderTests.ts:2422]: So are you saying the tests were not working? They were working a month ago... or did mocha release an update and break the tests... trying to understand as this project has a yarn.lock file

File: src/InMemoryProvider.ts (hunk 3 above)
Comment (Caseyyu2) [reply src/InMemoryProvider.ts:385]: updated.

File: src/tests/NoSqlProviderTests.ts (hunk 1 above)
Comment (Caseyyu2) [reply src/tests/NoSqlProviderTests.ts:2422]: According to the article, yes, it's a new release of mocha that broke the way it's checking done in an async function.
I'm not sure how the unit test was able to run before. But I need to both update the test file directory location in package.json and update the signature of the test files, then I was able to run.

File: src/tests/NoSqlProviderTests.ts (hunk 2 above)
Comment (Caseyyu2) [reply src/tests/NoSqlProviderTests.ts]: done

File: src/tests/NoSqlProviderTests.ts (hunk 1 above)
Comment (Caseyyu2) [reply src/tests/NoSqlProviderTests.ts:2422]: Discussed offline, this is to support higher version of mocha

File: src/InMemoryProvider.ts [hunk 6]
```diff
@@ -369,6 +369,21 @@ class InMemoryIndex extends DbIndexFTSFromRangeQueries {
...
+        });
+        if (isError(joinedKeys)) {
+            return Promise.reject(joinedKeys);
+        }
+
+        let values = [] as ItemType[];
+        for (const key of joinedKeys) {
+            values = values.concat(get(key, this._rbIndex) as ItemType[]);
+        }
+        return Promise.resolve(values && values[0] === undefined ? [] : values);
```
Comment (masterkidan) [inline src/InMemoryProvider.ts:384]: what would it happen when some of the keys were not found ... 
I think the earlier behavior is if you had a key value map like so 
1: foo
2: bar

And did a getMultiple(1, 2, 3) => would return [foo, bar, undefined] ... I don't think you should check for [0] alone to be undefined here.

File: src/IndexedDbProvider.ts [hunk 7]
```diff
@@ -755,6 +754,34 @@ class IndexedDbIndex extends DbIndexFTSFromRangeQueries {
...
+            for (const key of keys) {
+                values =  values.concat(await IndexedDbProvider.WrapRequest(this._store.get(key)));
+            }
+        } else {
+            for (const key of keys) {
+                values = values.concat(await this.getOnly(key));
+            }
+        }
+
+        return Promise.resolve(values && values[0] === undefined ? [] : values);
```
Comment (masterkidan) [inline src/IndexedDbProvider.ts:783]: Do you need to remove fullTextMetadata before returning here?

File: src/IndexedDbProvider.ts [hunk 8]
```diff
@@ -755,6 +754,34 @@ class IndexedDbIndex extends DbIndexFTSFromRangeQueries {
         return this._resolveCursorResult(req, limit, offset);
     }
 
+    async getMultiple(keyOrKeys: KeyType | KeyType[])
```
Comment (masterkidan) [inline src/IndexedDbProvider.ts:757]: Please remove the async here, see below as to why

File: src/IndexedDbProvider.ts [hunk 9]
```diff
@@ -755,6 +754,34 @@ class IndexedDbIndex extends DbIndexFTSFromRangeQueries {
...
+        if (isError(keys)) {
+            return Promise.reject(keys);
+        }
+
+        let values = [] as ItemType[];
+        if (this._store.getAll && !this._fakeComplicatedKeys) {
+            for (const key of keys) {
+                values =  values.concat(await IndexedDbProvider.WrapRequest(this._store.get(key)));
+            }
+        } else {
```
Comment (masterkidan) [inline src/IndexedDbProvider.ts:777]: Why the if, this doesn't make sense to me, can you explain? It looks like you are doing a _store.get either ways

File: src/IndexedDbProvider.ts [hunk 10]
```diff
@@ -755,6 +754,34 @@ class IndexedDbIndex extends DbIndexFTSFromRangeQueries {
...
+            for (const key of keys) {
+                values = values.concat(await this.getOnly(key));
```
Comment (masterkidan) [inline src/IndexedDbProvider.ts:779]: This would mean you would wait for 1 request to return before requesting a subsequent item, why not skip the await here and do a Promise.all() in the end.

File: src/NoSqlProvider.ts [hunk 11]
```diff
@@ -169,7 +170,14 @@ export abstract class DbProvider {
...
     }
 
-    getMultiple(storeName: string, keyOrKeys: KeyType | KeyType[]): Promise<ItemType[]> {
+    getMultiple(storeName: string, keyOrKeys: KeyType | KeyType[], indexName?: string): Promise<ItemType[]> {
+
+        if (indexName) {
+            return this._getStoreIndexTransaction(storeName, false, indexName).then(index => {
+                return index.getMultiple(keyOrKeys);
+            });
+        }
```
Comment (masterkidan) [inline src/NoSqlProvider.ts:179]: This block is unnecessary , getStoreIndexTransaction already handles the case when indexName is undefined

File: src/tests/NoSqlProviderTests.ts (hunk 2 above)
Comment (masterkidan) [reply src/tests/NoSqlProviderTests.ts]: Hmm, I guess my comment wasn't clear. 
i meant it should be a separate test that covers the following DataManipulation tests:

Simple index put/get, getAll, getOnly, and getRange‣
Simple index put/get, getAll, getOnly, and getRange (includeData)‣
Multipart primary key basic test‣
Multipart index basic test‣
Compound primary key basic test‣
Compound index basic test‣
MultiEntry multipart indexed tests‣
MultiEntry multipart indexed tests (includeData)‣
MultiEntry multipart indexed - update index‣
MultiEntry multipart indexed tests - Compound Key‣
MultiEntry multipart indexed - update index - Compound

You can do this by adding it to the tester() function in line 175

Also modify the first 2 specs in the above list to now include getMultiple as well.. so they would be 
simple index put/get, getAll, getOnly, getRange and getMultiple

File: src/InMemoryProvider.ts [hunk 12]
```diff
@@ -369,6 +369,21 @@ class InMemoryIndex extends DbIndexFTSFromRangeQueries {
...
+        const joinedKeys = attempt(() => {
+            return formListOfSerializedKeys(keyOrKeys, this._keyPath);
+        });
+        if (isError(joinedKeys)) {
+            return Promise.reject(joinedKeys);
+        }
+
+        let values = [] as ItemType[];
+        for (const key of joinedKeys) {
+            values = values.concat(get(key, this._rbIndex) as ItemType[]);
```
[the remark this answers is not in the document]
Comment (Caseyyu2) [reply src/InMemoryProvider.ts:382]: I agree the scenario mentioned above is a uncommon scenario. But this implementation supports scenario like people service calling getMultiple to get multiple users based on email address (non-primary keys). So joined keys with be different email address, and we are iterating through them. The implementation is on bar with the existing getMultiple API in inMemory provider for primary key.

File: src/InMemoryProvider.ts (hunk 6 above)
Comment (Caseyyu2) [reply src/InMemoryProvider.ts:384]: I debated about this too. The reason why I made the change is I was thinking most usage when consumer getting the result. Most likely they will try to iterate through the array, and accessing the property since [undefined] will be returned, and that will be an exception.

File: src/NoSqlProvider.ts (hunk 11 above)
Comment (Caseyyu2) [reply src/NoSqlProvider.ts:179]: Updated

File: src/IndexedDbProvider.ts (hunk 10 above)
Comment (Caseyyu2) [reply src/IndexedDbProvider.ts:779]: Good point. That's not ideal, I'll update that.

File: src/IndexedDbProvider.ts (hunk 9 above)
Comment (Caseyyu2) [reply src/IndexedDbProvider.ts:777]: Then logic for indexedDb when fakeComplicatedKeys is slightly different. so fakeComplicatedKeys is true in this case, if I call this._store.get(key) directly, the res is an object with only the keys and the refKeys
image
Only I call getOnly, it will go down the path try to call openCursor, I was able to get the full object back



I don't have too much context of what's the specific scenarios that fakeComplicatedKeys are set to true for creating indexedDB, but you can see all the other APIs have same support as well

File: src/IndexedDbProvider.ts (hunk 8 above)
Comment (Caseyyu2) [reply src/IndexedDbProvider.ts:757]: updated

File: src/IndexedDbProvider.ts (hunk 7 above)
Comment (Caseyyu2) [reply src/IndexedDbProvider.ts]: I have not seen indexedDB api use this function to nuke index. What scenario we need to do that?

File: src/InMemoryProvider.ts (hunk 6 above)
Comment (masterkidan) [reply src/InMemoryProvider.ts]: Lets remove this ... i think this check will only cause confusion... either we should loop through the array and remove all undefined values, or we should just leave them be... the later is better for perf

File: src/InMemoryProvider.ts [hunk 13]
```diff
@@ -369,6 +369,21 @@ class InMemoryIndex extends DbIndexFTSFromRangeQueries {
...
+        }
+        return Promise.resolve(compact(values));
```
Comment (masterkidan) [inline src/InMemoryProvider.ts:384]: wouldn't flatten be more appropriate to use here

File: src/IndexedDbProvider.ts [hunk 14]
```diff
@@ -755,6 +754,27 @@ class IndexedDbIndex extends DbIndexFTSFromRangeQueries {
...
+        });
+        if (isError(keys)) {
+            return Promise.reject(keys);
+        }
+
+        if (this._store.get && !this._fakeComplicatedKeys) {
+            return Promise.all<object[]>(map(keys, key => IndexedDbProvider.WrapRequest(this._store.get(key)))).then(compact);
+        }
+
+        return Promise.all(map(keys, key => this.getOnly(key))).then(vals => compact(flatten(vals)));
```
Comment (masterkidan) [inline src/IndexedDbProvider.ts:776]: Why not just call getOnly in both cases?
Also please call removeFullTextMetadataAndReturn() on the returned value... it is required to avoid returning metadata that is stored along with the data.

Comment (masterkidan) [inline src/IndexedDbProvider.ts:776]: why do you need to do both compact + flatten?

**Review by masterkidan — APPROVED** [review]
